### PR TITLE
Make chosing filename the job of the fetch_method

### PR DESCRIPTION
**Pull request by oxinabox**
Closes https://github.com/oxinabox/DataDeps.jl/issues/53
Which to quote:

> Because local file name is protocol dependent.
> And changing fetch method is how we change protocols

Basically, when using 
 - HTTP sometimes the filename in in the header, further ideally you would only get that header once, during get rather than first doing Head then doing Get
 - When using (For example, assuming we had a client) FTP, the filename comes from the `basename` of the URL
 - When using HTTP with Auth, then you need to send that auth header, in order to get back the filename header.

This also supersedes #45  because it solves that problem by not using string processing on URLs but by using the specified method.

The core change is that the `fetch_method` now takes a local directorypath, rather than a local filepath.
And then works out the filepath according to it's rules.
This is technically very breaking for anyone who is overloading the `fetch_method`.

Kinda coincidentally, I already opened a PR to HTTP last week (before I was consciously thinking about this)
https://github.com/JuliaWeb/HTTP.jl/pull/273
Once that is merged  all of `fetch_helpers` can be deleted, in favour of a thin (or zero) wrapper around that method.

@yakir12  This solves all the warnings in 
https://github.com/oxinabox/PyDrive.jl/blob/master/source/proto.ipynb
that I mentioned in
https://discourse.julialang.org/t/datadeps-and-google-drive/12271/11?u=oxinabox
Though that code needs to be updated to to return the filename.
Which is actually a trivial change since the filename is exposed as `metadata(gdfile, "title")`

@Evizero  do you think you might have time to do a code review on this at some-point?
It is I guess a pretty big change

### CODE REVIEW

**Comment by Evizero** [comment]
sure, can you request a review so I have it saved in my todo list?

**Comment by Evizero** [comment]
Thanks. For the record, I did not know a review-request requires collaborator status

File: src/fetch_helpers.jl [hunk 1]
```diff
@@ -1,5 +1,40 @@
 # This file is a part of DataDeps.jl. License is MIT.
 
+# TODO Remove this whole thing once https://github.com/JuliaWeb/HTTP.jl/pull/273
+
+"""
+    fetch_http(remotepath, localdir)
+
+Pass in a HTTP[/S] URL  and a directory to save it to,
+and it downloads that file, returing the local path.
+This using the HTTP protocol's method of defining filenames in headers,
```
Comment (Evizero) [inline src/fetch_helpers.jl]: "This **is** using ..."

**Comment by Evizero** [comment]
I browsed over MLDatasets to get a sense of what needs updating.

- The DataDeps themselves look fine, as none overwrite `fetch_method`. 

- Does anything change for `DataDeps.download` in the background? Its used directly twice in this file: https://github.com/JuliaML/MLDatasets.jl/blob/master/src/download.jl#L24

Either way, this looks like a good change.

**Comment by oxinabox** [comment]
> Does anything change for DataDeps.download in the background? Its used directly twice in this file: https://github.com/JuliaML/MLDatasets.jl/blob/master/src/download.jl#L24

Nothing observable, as 
`download(::DataDep,::String)` is a different higher level method, that eventually calls `fetch_method`
It is just that the default `fetch_method` is no longer `download(::String, ::String)`.

**Comment by oxinabox** [comment]
I'll fix that typo you found,
update the readme (one day I need to switch to using real docs, so I don't have as much copy paste between docstring and manual in the readme)

And debug A.V.'s failure (probably just it flaking out. I find AV has terrible network troubling)

**Pull request merged**